# Problem Identification

Personal financial **risk assessment is important to analyze because it allows us to reduce financial risk through effective risk management.**

Thus, this project is used to `analyze personal risk assessment` globally

The main objectives of this project is described using SMART Framework defined below.

> To reduce the occurrence of high financial risk ratings globally by 3% within 3 months, I will analyze a dataset gathered from Kaggle. This analysis will involve formulating relevant questions and creating visualizations based on the identified issues. The findings will help understand the contributing factors and provide targeted recommendations to address the high financial risk ratings.

As stated in my previous objectives, I want to analyze my dataset to answer the following questions:

1. What's the `sum of loan based on education level and risk rating`

2. What's the `average of debt-to-income ratio grouped by employment status and age group`

3. What's the `average assets based on their educational level and risk rating`

4. What's the `proportion of high financial risk rating by age group`

5. What's the `proportion of loan purposes by high financial risk rating`

6. What's the `distribution of debt-to-income ratio by financial risk rating`

7. How are the `education level and risk rating correlated?`

Based on my previous questions, I will use relevant data to support my analysis, which are:

- Age
- Assets Value
- Debt-to-Income Ratio
- Education Level
- Employment Status
- Income
- Loan Amount
- Loan Purpose
- Risk Rating

# Import Libs

In [1]:
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns
import seaborn.objects as so

import scipy.stats as stats

# Data Loading

In this section I will load the dataset that was gathered from [kaggle](https://www.kaggle.com/datasets/preethamgouda/financial-risk) and do some simple inspections, such as:

- Inspect the dataset's overview
- Inspect dataset's summary
- Inspect dataset's column names
- Inspect unique values of multiple columns in the dataframe

In [2]:
# define const
FILE_PATH = "dataset/financial_risk_assessment.csv"

In [3]:
# load dataset 
df = pd.read_csv(FILE_PATH)
df

Age      Gender Education Level Marital Status    Income  Credit Score  \
0       49        Male             PhD       Divorced   72799.0         688.0   
1       57      Female      Bachelor's        Widowed       NaN         690.0   
2       21  Non-binary        Master's         Single   55687.0         600.0   
3       59        Male      Bachelor's         Single   26508.0         622.0   
4       25  Non-binary      Bachelor's        Widowed   49427.0         766.0   
...    ...         ...             ...            ...       ...           ...   
14995   23  Non-binary      Bachelor's        Widowed   48088.0         609.0   
14996   56        Male             PhD         Single  107193.0         700.0   
14997   29  Non-binary             PhD        Married   46250.0         642.0   
14998   53  Non-binary             PhD       Divorced   40180.0         638.0   
14999   24  Non-binary      Bachelor's        Widowed       NaN         765.0   

       Loan Amount Loan Purpose Employment Status  Years at Current Job  \
0          45713.0     Business        Unemployed                    19   
1          33835.0         Auto          Employed                     6   
2          36623.0         Home          Employed                     8   
3          26541.0     Personal        Unemployed                     2   
4          36528.0     Personal        Unemployed                    10   
...            ...          ...               ...                   ...   
14995      26187.0         Home     Self-employed                     2   
14996      35111.0         Auto     Self-employed                    10   
14997      44369.0         Home        Unemployed                    19   
14998      32752.0         Home     Self-employed                    12   
14999          NaN     Personal     Self-employed                    18   

      Payment History  Debt-to-Income Ratio  Assets Value  \
0                Poor              0.154313      120228.0   
1                Fair              0.148920       55849.0   
2                Fair              0.362398      180700.0   
3           Excellent              0.454964      157319.0   
4                Fair              0.143242      287140.0   
...               ...                   ...           ...   
14995            Fair              0.317633           NaN   
14996            Fair              0.155126       79102.0   
14997       Excellent              0.593999      196930.0   
14998       Excellent              0.478035      276060.0   
14999       Excellent              0.116083       71699.0   

       Number of Dependents                 City State  \
0                       0.0       Port Elizabeth    AS   
1                       0.0      North Catherine    OH   
2                       3.0          South Scott    OK   
3                       3.0           Robinhaven    PR   
4                       NaN          New Heather    IL   
...                     ...                  ...   ...   
14995                   4.0            Susanstad    TN   
14996                   NaN         Port Heather    WA   
14997                   4.0  South Morganchester    LA   
14998                   NaN           Port Wayne    AK   
14999                   3.0          South Stacy    WA   

                         Country  Previous Defaults  Marital Status Change  \
0                         Cyprus                2.0                      2   
1                   Turkmenistan                3.0                      2   
2                     Luxembourg                3.0                      2   
3                         Uganda                4.0                      2   
4                        Namibia                3.0                      1   
...                          ...                ...                    ...   
14995                   Djibouti                2.0                      0   
14996                      Congo                0.0                      0   
1

Insights:
- There are 15,000 rows and 20 columns in the dataset
- There are some missing values on assets value and number of dependents, but there might be any on other columns as well

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Age                    15000 non-null  int64  
 1   Gender                 15000 non-null  object 
 2   Education Level        15000 non-null  object 
 3   Marital Status         15000 non-null  object 
 4   Income                 12750 non-null  float64
 5   Credit Score           12750 non-null  float64
 6   Loan Amount            12750 non-null  float64
 7   Loan Purpose           15000 non-null  object 
 8   Employment Status      15000 non-null  object 
 9   Years at Current Job   15000 non-null  int64  
 10  Payment History        15000 non-null  object 
 11  Debt-to-Income Ratio   15000 non-null  float64
 12  Assets Value           12750 non-null  float64
 13  Number of Dependents   12750 non-null  float64
 14  City                   15000 non-null  object 
 15  St

In [5]:
df.isna().sum()

Age                         0
Gender                      0
Education Level             0
Marital Status              0
Income                   2250
Credit Score             2250
Loan Amount              2250
Loan Purpose                0
Employment Status           0
Years at Current Job        0
Payment History             0
Debt-to-Income Ratio        0
Assets Value             2250
Number of Dependents     2250
City                        0
State                       0
Country                     0
Previous Defaults        2250
Marital Status Change       0
Risk Rating                 0
dtype: int64

## Insights & Actions

Insights:   
- There are 15,000 observations and 20 columns in the dataset.
- There are some columns containing missing values of 2,250 observations, such as: `Income`, `Credit Score`, `Loan Amount`, `Assets Value`, `Number of Dependents`

- Some columns are using inappropriate data type, this problem defined as follows:
    - Should be `categorical`:
        - Employment Status
        - Education Level
        - Risk Rating
        - Loan Purpose

Actions:
- Adjust data type of columns defined above
- Identify missing values whether the are MAR, MCAR, or MNAR. But, **I need to filter required columns first**

In [6]:
df.columns

Index(['Age', 'Gender', 'Education Level', 'Marital Status', 'Income',
       'Credit Score', 'Loan Amount', 'Loan Purpose', 'Employment Status',
       'Years at Current Job', 'Payment History', 'Debt-to-Income Ratio',
       'Assets Value', 'Number of Dependents', 'City', 'State', 'Country',
       'Previous Defaults', 'Marital Status Change', 'Risk Rating'],
      dtype='object')

Insights:

- All columns' names are clear and ready to be used for further analysis

In [7]:
# loop through columns with object data type
for col in df.select_dtypes(include="object").columns:
        print(col)
        print(df[col].unique(), end="\n\n")

Gender
['Male' 'Female' 'Non-binary']

Education Level
['PhD' "Bachelor's" "Master's" 'High School']

Marital Status
['Divorced' 'Widowed' 'Single' 'Married']

Loan Purpose
['Business' 'Auto' 'Home' 'Personal']

Employment Status
['Unemployed' 'Employed' 'Self-employed']

Payment History
['Poor' 'Fair' 'Excellent' 'Good']

City
['Port Elizabeth' 'North Catherine' 'South Scott' ... 'Susanstad'
 'South Morganchester' 'Port Wayne']

State
['AS' 'OH' 'OK' 'PR' 'IL' 'TN' 'MD' 'MA' 'DC' 'VT' 'NH' 'NM' 'IA' 'AZ'
 'MH' 'MO' 'MS' 'MP' 'DE' 'UT' 'KS' 'VI' 'CO' 'OR' 'VA' 'GA' 'ID' 'HI'
 'AL' 'TX' 'WY' 'KY' 'WA' 'RI' 'LA' 'MI' 'WV' 'ND' 'NY' 'NJ' 'CT' 'PW'
 'NV' 'FM' 'CA' 'NC' 'IN' 'GU' 'AR' 'WI' 'SD' 'MT' 'MN' 'NE' 'SC' 'ME'
 'AK' 'PA' 'FL']

Country
['Cyprus' 'Turkmenistan' 'Luxembourg' 'Uganda' 'Namibia' 'Iceland'
 'Bouvet Island (Bouvetoya)' 'Honduras' 'Pitcairn Islands' 'Thailand'
 'French Guiana' 'Antarctica (the territory South of 60 deg S)' 'Eritrea'
 'Grenada' 'Solomon Islands' 'Guadeloup

Insights:
- Gender, Eduational Level, Marital Status, Loan Purpose, Employment Status, Payment History, City, State, Country, and Risk Rating columns are categorical columns and need a data type adjustment

# Data Cleaning & Preparation

This section used to prepare and clean the dataset for further analysis

In [8]:
# create a copy of dataframe to preserve the original df
df_clean = df.copy()

## Filter Required Columns

As mentioned earlier, the columns that needed for further analysis are:
- Age
- Assets Value
- Debt-to-Income Ratio
- Education Level
- Employment Status
- Income
- Loan Amount
- Loan Purpose
- Risk Rating

In [9]:
# filter columns
df_clean = df_clean[[
        "Age",
        "Assets Value",
        "Debt-to-Income Ratio",
        "Education Level",
        "Employment Status",
        "Income",
        "Loan Amount",
        "Loan Purpose",
        "Risk Rating",
]].copy()

df_clean.head()

Age  Assets Value  Debt-to-Income Ratio Education Level Employment Status  \
0   49      120228.0              0.154313             PhD        Unemployed   
1   57       55849.0              0.148920      Bachelor's          Employed   
2   21      180700.0              0.362398        Master's          Employed   
3   59      157319.0              0.454964      Bachelor's        Unemployed   
4   25      287140.0              0.143242      Bachelor's        Unemployed   

    Income  Loan Amount Loan Purpose Risk Rating  
0  72799.0      45713.0     Business         Low  
1      NaN      33835.0         Auto      Medium  
2  55687.0      36623.0         Home      Medium  
3  26508.0      26541.0     Personal      Medium  
4  49427.0      36528.0     Personal         Low

> Successfully filter needed columns

## Create New Columns

In this section, I will create new columns to support my analysis:
- Since `I need the age to be groupped` and the dataset contains numerical age, I will `create new column called Age Group`

In [10]:
# use cut() to group age
df_clean["Age Group"] = pd.cut(
    df_clean['Age'], 
    bins=[18, 28, 38, 48, 58, 69], 
    labels=["18-28", "28-38", "38-48", "48-58", "58-69"], 
    right=True,
    include_lowest=True
)

In [11]:
df_clean.head()

Age  Assets Value  Debt-to-Income Ratio Education Level Employment Status  \
0   49      120228.0              0.154313             PhD        Unemployed   
1   57       55849.0              0.148920      Bachelor's          Employed   
2   21      180700.0              0.362398        Master's          Employed   
3   59      157319.0              0.454964      Bachelor's        Unemployed   
4   25      287140.0              0.143242      Bachelor's        Unemployed   

    Income  Loan Amount Loan Purpose Risk Rating Age Group  
0  72799.0      45713.0     Business         Low     48-58  
1      NaN      33835.0         Auto      Medium     48-58  
2  55687.0      36623.0         Home      Medium     18-28  
3  26508.0      26541.0     Personal      Medium     58-69  
4  49427.0      36528.0     Personal         Low     18-28

In [12]:
df_clean.isna().sum()

Age                        0
Assets Value            2250
Debt-to-Income Ratio       0
Education Level            0
Employment Status          0
Income                  2250
Loan Amount             2250
Loan Purpose               0
Risk Rating                0
Age Group                  0
dtype: int64

## Adjust Column Data Type

In [14]:
df_clean.dtypes

Age                        int64
Assets Value             float64
Debt-to-Income Ratio     float64
Education Level           object
Employment Status         object
Income                   float64
Loan Amount              float64
Loan Purpose              object
Risk Rating               object
Age Group               category
dtype: object

In this part, I will adjust columns' data types based on following scenarios:
- Change to `category`
    - Employment Status
    - Education Level (`ordinal`)
    - Risk Rating (`ordinal`)
    - Loan Purpose
 
- Change to `int`
    - Assets Value
    - Income
    - Loan Amount

### Ordinal

In [17]:
def dTypeToOrdinal(series:str, cat_order_list:list):
    """
    Adjust :series data type into ordinal category based on :cat_order_list

    :series -> str
    :cat_order_list -> list

    :return series with ordinal data type
    """
    
    return pd.Categorical(
        series, 
        categories=cat_order_list, 
        ordered=True
    )

In [18]:
education_level_cat_order = ["High School", "Bachelor's", "Master's", "PhD"]
df_clean["Education Level"] = dTypeToOrdinal(
    df_clean["Education Level"],
    education_level_cat_order
)

In [19]:
df_clean["Education Level"].dtype

CategoricalDtype(categories=['High School', 'Bachelor's', 'Master's', 'PhD'], ordered=True, categories_dtype=object)

In [20]:
risk_rating_cat_order = ["Low", "Medium", "High"]
df_clean["Risk Rating"] = dTypeToOrdinal(
    df_clean["Risk Rating"],
    risk_rating_cat_order
)

In [21]:
df_clean["Risk Rating"].dtype

CategoricalDtype(categories=['Low', 'Medium', 'High'], ordered=True, categories_dtype=object)

### Nominal

In [22]:
# convert these columns into category
df_clean[[
    "Employment Status", 
    "Loan Purpose"
]] = df_clean[[
    "Employment Status", 
    "Loan Purpose"
]].astype("category")

In [23]:
# check columns' data types
df_clean.dtypes

Age                        int64
Assets Value             float64
Debt-to-Income Ratio     float64
Education Level         category
Employment Status       category
Income                   float64
Loan Amount              float64
Loan Purpose            category
Risk Rating             category
Age Group               category
dtype: object

### Int

In [24]:
# change data types to int and ignore missing values to be converted into int
df_clean[[
    "Assets Value",
    "Income",
    "Loan Amount"
]] = df_clean[[
    "Assets Value",
    "Income",
    "Loan Amount"
]].astype("int", errors="ignore")

In [25]:
df_clean.dtypes

Age                        int64
Assets Value             float64
Debt-to-Income Ratio     float64
Education Level         category
Employment Status       category
Income                   float64
Loan Amount              float64
Loan Purpose            category
Risk Rating             category
Age Group               category
dtype: object

Insights:
- Since assets value, income, and loan amout columns contain Nan, they can't be converted into int

Actions:
- **Missing values must be dropped first** before converting the data type into int

## Missing Value Type Identification

In [26]:
df_clean.isna().sum()

Age                        0
Assets Value            2250
Debt-to-Income Ratio       0
Education Level            0
Employment Status          0
Income                  2250
Loan Amount             2250
Loan Purpose               0
Risk Rating                0
Age Group                  0
dtype: int64

> In this section, I will identify whether missing values in `Assets Value`, `Income`, and `Loan Amount` columns can be categorized as MAR, MCAR, or MNAR.

In [27]:
def get_freq_on_specific_cat_col(selected_col: str, df: pd.DataFrame):
    """
    Get frequency of each categories on :selected_col inside :df

    :selected_col as str
    :df as Dataframe

    :return series of value counts
    """
    return df[selected_col].value_counts()

In [28]:
def get_med_mean_on_specific_num_col(selected_col: str, df: pd.DataFrame):
    """
    Get mean and median of each numerical columns on :selected_col inside :df

    :selected_col as str
    :df as Dataframe

    :return mean and med
    """
    return pd.DataFrame(df[selected_col].describe()[["50%", "mean"]].map(lambda val: f'{val:,.2f}'))

### Missing Values on Assets Value

In [29]:
# filter rows with missing values and non missing values on assets value
missing_assets_value = df_clean.loc[df_clean["Assets Value"].isna(), df_clean.columns!="Assets Value"]
non_missing_assets_value = df_clean.loc[df_clean["Assets Value"].notna(), df_clean.columns!="Assets Value"]

#### Among Categorical Column

In [30]:
# loop through category columns in missing_assets_value
for col_name in missing_assets_value.select_dtypes(include="category"):
    print("Missing Values Frequency of Assets on ")
    print(get_freq_on_specific_cat_col(col_name, missing_assets_value), end="\n\n")

Missing Values Frequency of Assets on 
Education Level
Master's       569
Bachelor's     565
PhD            563
High School    553
Name: count, dtype: int64

Missing Values Frequency of Assets on 
Employment Status
Employed         785
Self-employed    765
Unemployed       700
Name: count, dtype: int64

Missing Values Frequency of Assets on 
Loan Purpose
Home        597
Personal    558
Auto        555
Business    540
Name: count, dtype: int64

Missing Values Frequency of Assets on 
Risk Rating
Low       1369
Medium     641
High       240
Name: count, dtype: int64

Missing Values Frequency of Assets on 
Age Group
58-69    486
18-28    484
28-38    435
48-58    432
38-48    413
Name: count, dtype: int64



Insights:

As we can see that missing values frequency of `Assets Values` among all other categorical columns appear randomly. In other words, **the missing values are missing completely at random**

#### Among Numerical Column

To identify missing values type among all numerical columns, I will compare median and mean of all numericals columns of missing and non missing values in asset value column

In [32]:
# make sure all columns in missing_assets_value and non_missing_assets_value are the same
if (set(missing_assets_value.columns) == set(non_missing_assets_value.columns)):
    # loop through numerical columns
    for col_name in missing_assets_value.select_dtypes(include=["int", "float"]):
        print("Med & Mean: ")
        print("Missing Values on Asset Values: ")
        print(get_med_mean_on_specific_num_col(col_name, missing_assets_value))
        print("Non-missing values on Asset Values: ")
        print(get_med_mean_on_specific_num_col(col_name, non_missing_assets_value), end="\n\n")

Med & Mean: 
Missing Values on Asset Values: 
        Age
50%   43.00
mean  43.40
Non-missing values on Asset Values: 
        Age
50%   43.00
mean  43.46

Med & Mean: 
Missing Values on Asset Values: 
     Debt-to-Income Ratio
50%                  0.35
mean                 0.35
Non-missing values on Asset Values: 
     Debt-to-Income Ratio
50%                  0.35
mean                 0.35

Med & Mean: 
Missing Values on Asset Values: 
         Income
50%   69,703.00
mean  70,007.50
Non-missing values on Asset Values: 
         Income
50%   69,830.00
mean  69,920.25

Med & Mean: 
Missing Values on Asset Values: 
     Loan Amount
50%    27,734.50
mean   27,411.13
Non-missing values on Asset Values: 
     Loan Amount
50%    27,500.00
mean   27,456.93



Insights:

These values are inspected for **missing values and non-missing values** in `Asesst values` column

- Median of Age for missing values and non-missing values are both 43.
- Mean of Age for missing values and non-missing values are 43.40 and 43.46 respectively.

- Median and mean of Debt-to-Income ratio for missing values and non-missing values are all 0.35.

- Median of Income for missing values and non-missing values are \\$69,703.00 and \\$69,830.00 respectively
- Mean of Income for missing values and non-missing values are \\$70,007.50 and \\$69,920.25 respectively.

- Median of Loan Amount for missing values and non-missing values are \\$27,734.50 and \\$27,500.00 respectively
- Mean of Loan Amount for missing values and non-missing values are \\$27,411.13 and \\$27,456.93 respectively.


> Since the numerical column distribution in data with missing values and without missing values on assets value are very simmilar. I can conclude that the missing values is MCAR and safe to drop.

Actions:
- Drop rows with missing values on loan amount column

## Missing Values on Income

In [31]:
# filter rows with missing values and non missing values on income
missing_income = df_clean.loc[df_clean["Income"].isna(), df_clean.columns!="Income"]
non_missing_income = df_clean.loc[df_clean["Income"].notna(), df_clean.columns!="Income"]

In [34]:
# make sure all columns in missing_assets_value and non_missing_assets_value are the same
if (set(missing_income.columns) == set(non_missing_income.columns)):
    for col_name in missing_income.select_dtypes(include=["int", "float"]):
        print("Med & Mean: ")
        print("Missing Values on Income: ")
        print(get_med_mean_on_specific_num_col(col_name, missing_income))
        print("Non-missing values on Income: ")
        print(get_med_mean_on_specific_num_col(col_name, non_missing_income), end="\n\n")

Med & Mean: 
Missing Values on Income: 
        Age
50%   44.00
mean  43.89
Non-missing values on Income: 
        Age
50%   43.00
mean  43.37

Med & Mean: 
Missing Values on Income: 
     Assets Value
50%    157,070.00
mean   159,623.89
Non-missing values on Income: 
     Assets Value
50%    159,925.00
mean   159,762.36

Med & Mean: 
Missing Values on Income: 
     Debt-to-Income Ratio
50%                  0.35
mean                 0.35
Non-missing values on Income: 
     Debt-to-Income Ratio
50%                  0.35
mean                 0.35

Med & Mean: 
Missing Values on Income: 
     Loan Amount
50%    27,866.00
mean   27,466.52
Non-missing values on Income: 
     Loan Amount
50%    27,499.00
mean   27,447.10



## Missing Values on Loan Amount

In [ ]:
# filter rows with missing values and non missing values on loan_amount
missing_loan_amount = df_clean.loc[df_clean["Loan Amount"].isna(), df_clean.columns!="Loan Amount"]
non_missing_loan_amount = df_clean.loc[df_clean["Loan Amount"].notna(), df_clean.columns!="Loan Amount"]

#### Among Categorical Column

In [ ]:
# loop through category columns in missing_loan_column
for col_name in missing_loan_amount.select_dtypes(include="category"):
    print("Missing Values Frequency of Loan Amount on")
    print(missing_freq_on_cat_col(col_name, missing_loan_amount), end="\n\n")

Insights:

As we can see that missing values frequency of `Loan Amount` among all other categorical columns appear randomly. In other words I can assume that the missing values are missing completely at random

#### Among Numerical Column

##### Missing Loan Amount only

In [ ]:
# loop through numerical columns inside missing_loan_amount
for col_name in missing_loan_amount.select_dtypes(include=["int", "float"]):
    print("Desc Stats of {} for missing values in loan amount".format(col_name))
    print(missing_freq_on_num_col(col_name, missing_loan_amount), end="\n\n")

##### Non-Missing Loan Amount only

In [ ]:
# loop through numerical columns inside non_missing_loan_amount
for col_name in non_missing_loan_amount.select_dtypes(include=["int", "float"]):
    print("Desc Stats of {} for non-missing values in loan amount".format(col_name))
    print(missing_freq_on_num_col(col_name, non_missing_loan_amount), end="\n\n")

Insights:

- Descriptive statistics of missing values and non-missing values in loan amount among numerical columns (`Debt-to-Income Ratio`, `Age`, `Assets Value`) are almost similar.


- Median of Debt-to-Income ratio for missing values and non-missing values are 0.34 and 0.35 respectively. These values are very similar.

- Mean of Debt-to-Income ratio for missing values and non-missing values in loan amount are both 0.35. These values are identical.



- Median of Age for missing values and non-missing values are both 43. These values are identical.

- Mean of Age for missing values and non-missing values in loan amount are 43.05 and 43.52 respectively. These values are identical.



- Median of Assets Value for missing values and non-missing values in amount are $157,812.00 and $159,486.50 respectively. These values are very similar.

- Mean of Assets Value for missing values and non-missing values in loan amount is $157,898.22 and $160,069.49 respectively. These values are very similar.

> Since the numerical column distribution in data with missing values and without missing values on loan amount are very simmilar. I can conclude that the missing values is MCAR and safe to drop.

Actions:
- Drop rows with missing values on loan amount column

## Dropping Missing Values

Since all missing values in `Income` and `Loan Amount` are MCAR. I can delete all rows with missing values safely

In [ ]:
# drop any rows with missing values
df_clean = df_clean.dropna()

In [ ]:
# missing values inspection for each column
df_clean.isna().sum()

> All missing values are droped successfully

# Analysis

In this sections I will create analysis to answer my questions thruogh visualizations.

1. What's the `sum of loan based on education level and risk rating`

2. What's the `average of debt-to-income ratio grouped by employment status and age group`

3. What's the `average assets based on their educational level and risk rating`

4. What's the `proportion of high financial risk rating by age group`

5. What's the `proportion of loan purposes by high financial risk rating`

6. What's the `distribution of debt-to-income ratio by financial risk rating`

7. How are the education level and risk rating correlated?

## SUM of loan based on education level and risk rating

In [ ]:
# find sum of loan amount by education level and risk rating
df_clean.groupby(["Education Level", "Risk Rating"], observed=True).agg({"Loan Amount": "sum"})

In [ ]:
# barplot with title and custom y label and legend
sns.barplot(
    df_clean, 
    x="Education Level", 
    y="Loan Amount",
    hue="Risk Rating",
    estimator="sum",
    errorbar=None
)

plt.title("SUM of Loan Amount by Education Level and Financial Risk Rating")
plt.ylabel("SUM of Loan Amount")
plt.legend(loc="lower right", title="Financial Risk Rating")
plt.show()

Insights:   
- **Among all Education Levels, highest loan in total are considered as low financial risk rating.**
- Lowest sum of loan in total, precisely having high financial risk rating.
- This might be because **people with high amount of loan able to manage their financial and generate more revenue with it.**

## AVG of Debt-to-Income Ratio by Employment Status and Age Group

In [ ]:
# find debt to income ration by employment status and age group
df_clean.groupby(["Employment Status", "Age Group"], observed=True).agg({"Debt-to-Income Ratio": "mean"})

In [ ]:
# barplot with title and custom y label and legend
sns.barplot(
    df_clean, 
    x="Employment Status", 
    y="Debt-to-Income Ratio",
    hue="Age Group",
    estimator="mean",
    errorbar=None
)

plt.title("AVG of Debt-to-Income Ratio by Employment Status and Age Group")
plt.ylabel("AVG of Debt-to-Income Ratio")
plt.legend(loc="lower right", title="Age Group")
plt.show()

Insights:   
- Average of Debt-to-Income ratio (DTI) for each employment status and grouped by age group seem similar.
- Age group of 18-28 for self-employed are having the highest DTI on the group, but not significanly compared to others
- Employed people within range of age 48-58 are having highest DTI, also not significanly compared to others
- Based on the visualization **I conclude that Age and Employment Status don't have any correlation to DTI.**

In [ ]:
df_clean.columns

## AVG of assets based on their educational level and risk rating

In [ ]:
# find sum of assets value by education level and risk rating
df_clean.groupby(["Education Level", "Risk Rating"], observed=True).agg({"Assets Value": "sum"})

In [ ]:
# show barplot with label and custom ylabel & legend
sns.barplot(
    df_clean,
    y="Education Level",
    x="Assets Value",
    orient="h",
    hue="Risk Rating",
    estimator="sum",
    errorbar=None
)

plt.title("SUM of Assets by Education Level and Financial Risk Rating")
plt.ylabel("SUM of Assets")
plt.legend(loc="upper right", title="Financial Risk Rating")
plt.show()

Insights:
- The graph shown that people with highet assets are having low financial risk rating.
- People with lowest assets are having high financial risk rating.

## Proportion of High Financial Risk Rating by Age Group

In [ ]:
# filter high risk rating only and get age group column
prop_high_fin_risk_by_age_group = df_clean.loc[df_clean["Risk Rating"]=="High", "Age Group"].value_counts(normalize=True)
prop_high_fin_risk_by_age_group

In [ ]:
# show pie plot with title
plt.pie(prop_high_fin_risk_by_age_group, labels=prop_high_fin_risk_by_age_group.keys(), autopct='%.1f%%')

plt.title("Proportion of High Financial Risk Rating by Age Group")
plt.show()

Insights: 
- Age group 58-69 is the highest proportion of high financial risk rating by 22.64%.
- The proportion of high financial risk rating among all age group is very similar.
- In conclution, the graph shown that **age doesn't have correlation with financial risk, since the proportions are almost identical**

## Proportion of Loan Purposes by High Financial Risk Rating

In [ ]:
# filter high risk rating only and ge loan purpose column
prop_loan_purpose_by_high_fin_rating = df_clean.loc[df_clean["Risk Rating"]=="High", "Loan Purpose"].value_counts(normalize=True)
prop_loan_purpose_by_high_fin_rating

In [ ]:
# show pie plot with title
plt.pie(
    prop_loan_purpose_by_high_fin_rating, 
    labels=prop_loan_purpose_by_high_fin_rating.keys(), 
    autopct='%.1f%%'
)

plt.title("Proportion of Loan Purposes by High Financial Risk Rating")
plt.show()

Insights: 
- `Auto is the highest loan purpose from high financial risk rating` followed by personal and home.
- Proportion of loan purposes with high financial risk rating are almost equal.
- `All loan purposes have almost equal impact on High Financial Risk Rating`
- Hence, I conclude that **there's no assosiation between these variables**

## Debt-to-Income Ratio Distribution among Education Level

In [ ]:
sns.boxplot(
    df_clean, 
    x="Employment Status",
    y="Debt-to-Income Ratio"
)

Insights:
- DTI distributions among Employment Status are also exactly the same.
- Thus, **Employment Status and DTI don't have any assosiations**

## Debt-to-income Ratio Distribution among Financial Risk Rating

In [ ]:
# show boxplot of debt-to-income ratio on each risk rating with title
sns.boxplot(data=df_clean, x="Risk Rating", y="Debt-to-Income Ratio")
plt.title("Distribution of Debt-to-Income Ratio among Financial Risk Rating")
plt.show()

Insights:
- Data distribution of Debt-to-Income ratio among all Risk Rating categories are equal
- Hence, **there is no association between the Debt-to-Income ratio and Risk Rating categories**

## Debt-to-Income Ratio Distribution among Employment Status

In [ ]:
sns.boxplot(
    df_clean, 
    x="Education Level",
    y="Debt-to-Income Ratio"
)

Insights:
- DTI distributions among education level are exactly the same
- Therefore, Education Level doesn't have assosiations with DTI

## srtsrt

In [ ]:
df_clean.columns

In [ ]:
# new columns to contain loan to income ratio range values: 0-1
df_clean["Loan-to-Income Ratio"] = df_clean["Loan Amount"]/df_clean["Income"]

In [ ]:
df_clean[["Income", "Loan Amount", "Loan-to-Income Ratio"]]

In [ ]:
# show barplot with label and custom ylabel & legend
(
    so.Plot(
        df_clean,
        y="Education Level",
        x="Assets Value",
        color="Risk Rating",
    ).add(so.Bar(), so.Count(), so.Stack())
)

## How are the education level and risk rating correlated?

I will answer the question by conducting hypothesis testing stated below.
- H0: There's no correlation between education level and risk rating (The variables are independent)
- H1: Education level and risk rating are correlated to each other (The variables are not independent)

In [ ]:
# crosstab on education level and risk rating
education_level_vs_risk_rating = pd.crosstab(df_clean["Education Level"], df_clean["Risk Rating"])
education_level_vs_risk_rating

In [ ]:
# chi squared test
"Pval: {:.2f}".format(stats.chi2_contingency(education_level_vs_risk_rating).pvalue)

Insights:
- Since the p-value is greater that critical value (0.53 > 0.05), the hypothesis test shown that H0 is fail to rejected.

- Which means that **there's no significant evidence that education level and risk rating are having an assosiation**

# Conclutions

- People who have **high total loan amount and assets tend to have a low financial risk rating.**
- The Debt-to-Income Ratio (DTI) is almost identical across all age groups, regardless of whether individuals are unemployed, self-employed, or employed.
- All loan purposes are having nearly identical proportion on High Financial Risk Rating
- All Age groups categories are also having nearly identical proportion on Financial Risk Ratings categories
- **There's no assotiations between DTI and Financial Risk Ratings categories**
- **There's no significant assotiations between Education Level and Financial Risk Rating**

> Individuals needs to increase their total assets and learn to manage their loan, so they can generate more revenue which can lead to decrease their financial risk rating.

# Export Clean Data

In [ ]:
# export cleaned dataset into .csv format
# df_clean.to_csv("dataset/clean_financial_risk_assesment.csv", index=False)